### ISSUE TITLE
Lost in the Java Jungle (TM)

### ISSUE THREAD

**Issue by borismarin** [tarelli/org.lemsml.model#19]
I've been playing with java libraries (packages?) for the unit parsing part and expression evaluation. There seems to be a huge library that can do both, called JScience -- which has even been [investigated by](https://github.com/tarelli/jscience) @tarelli in the past. The library seems a little bit dormant, though. The unit handling part of it seems to have been "incorporated" into javax (does that mean it ships with the jvm?). But that thing is associated with something called JSR-275 which [has been rejected](https://groups.google.com/forum/#!topic/javaposse/4kisKkrIuVI).  The guy who wrote JScience seems to be involved with something else called JSR-363, which is pretty much [alive](https://github.com/unitsofmeasurement).  And it seems to be related to [this eclipse thingy](http://www.eclipse.org/uomo/). As you can see, there are lots of "things" which I can't even classify, forgive my ignorance.

In summary, how do you guys usually decide on those things? What should be considered "reputable/robust/safe"?

**Comment by borismarin** [comment]
There seems to be some standardization discussion regarding units: http://unitsofmeasure.org/trac/

**Comment by robertcannon** [comment]
No idea about robust/safe, but the uom-se package looks pretty good. If you dig down far enough you find
 https://github.com/unitsofmeasurement/uom-se/blob/master/src/main/java/tec/uom/se/quantity/QuantityDimension.java   which is pretty much an exact parallel of the Dimension class in jLEMS.

If this is going to become a standard part of Java 8, then it sounds like a pretty good choice. If GWT decides to support it one day, I could even use it for jLEMSDev.

**Comment by tarelli** [comment]
@borismarin we are using that in Geppetto look at this bit of code https://github.com/openworm/org.geppetto.simulator.jlems/blob/master/src/main/java/org/geppetto/simulator/jlems/JLEMSSimulatorService.java#L456

**Comment by borismarin** [comment]
Lovely. Will blindly steal it.

**Comment by borismarin** [comment]
BTW, @tarelli, I couldn't steal that blindly because it's doing sthing different (and wrong, which seems to be the reason your are [not using it](https://github.com/openworm/org.geppetto.simulator.jlems/blob/master/src/main/java/org/geppetto/simulator/jlems/JLEMSSimulatorService.java#L329).). I just had a quick look at it, but I guess (one of the) reason(s) it's not working is the fact that you are mixing scaling factors (like pico, nano) with base unit powers. In other words, if what you have in `String dimension` is indeed [a comma-separated list of dimension powers ...](https://github.com/openworm/org.geppetto.simulator.jlems/blob/master/src/main/java/org/geppetto/simulator/jlems/JLEMSSimulatorService.java#L458), you will need to  `getUnit(x, SI.X).pow(Integer.parseInt(st.nextToken())` instead of what is in there.

**Comment by borismarin** [comment]
Decided to stick with http://unitsofmeasurement.github.io/ . It seems to concentrate all previous efforts, and will hopefully be accepted into javax.measure once it's ready. Some quick and dirty tests/examples [here](https://github.com/borismarin/lems-expr-parser/commit/2be8d38b81a63f5b1b393e93802afadf0eaf961c).
